lib/tar.ml

```diff
@@ -34,28 +34,32 @@ module Header = struct
     String.(trim (map (function '\000' -> ' ' | x -> x) s))
 
   (** Unmarshal an integer field (stored as 0-padded octal) *)
-  let unmarshal_int x =
-    let tmp = "0o0" ^ (trim_numerical x) in
+  let unmarshal_int ~off ~len x =
+    let tmp = "0o0" ^ (trim_numerical (String.sub x off len)) in
     try
       Ok (int_of_string tmp)
     with Failure msg ->
       Error (`Unmarshal (Printf.sprintf "%s: failed to parse integer %S" msg tmp))
 
   (** Unmarshal an int64 field (stored as 0-padded octal) *)
-  let unmarshal_int64 x =
-    let tmp = "0o0" ^ (trim_numerical x) in
+  let unmarshal_int64 ~off ~len x =
+    let tmp = "0o0" ^ (trim_numerical (String.sub x off len)) in
     try
       Ok (Int64.of_string tmp)
     with Failure msg ->
       Error (`Unmarshal (Printf.sprintf "%s: failed to parse int64 %S" msg tmp))
 
   (** Unmarshal a string *)
-  let unmarshal_string x =
+  let unmarshal_string ?(off = 0) ?len x =
+    let len = Option.value ~default:(String.length x - off) len in
     try
-      let first_0 = String.index x '\000' in
-      Ok (String.sub x 0 first_0)
+      let first_0 = String.index_from x off '\000' in
```

Comment:
`String‧index_from` returns an absolute offset instead of a relative one to the given `off`. In other words, we must do `first_0 - off` then for the next `if` and the next `String‧sub`.